tools/ai2_common.py

```diff
@@ -0,0 +1,230 @@
+        self.arg2 = parent_doc.get_entity(brat_annotation.arg2)
+
+
+def get_docs(*paths):
+    found_paths = []
+    extensions = re.compile(r"\.(txt|ann)$")
+    for path in paths:
+        if os.path.isdir(path):
+            # Don't recur, just check one level down
+            children = [f for f in glob.glob(os.path.join(path, "*")) if extensions.search(f)]
+            if not children:
+                sys.stderr.write("No annotation files found in {}\n".format(path))
+            identifiers = [child[:-4] for child in children]
+            found_paths += list(set(identifiers))
+        else:
+            if extensions.search(path):
+                found_paths.append(path[:-4])
+            else:
+                found_paths.append(path)
+
+    result = [load_doc(p) for p in found_paths]
+    return result
+
+
+def load_doc(identifier):
+    """An identifier is the path to a pair of a .ann file and a .txt file, minus
+    the extension for either.
+
+    """
+    return EnhancedAnnotatedDoc(annotation.TextAnnotations(identifier))
+
+
+def get_token_starting_at_char_offset(doc, offset):
+    l = 0
+    r = len(doc)
+    while r - l > 1:
+        mid = (l + r) / 2
+        if doc[mid].idx > offset:
+            r = mid
+        else:
+            l = mid
+
+    if doc[l].idx == offset:
+        return doc[l]
+    else:
+        return None
+
+
+def get_token_ending_at_char_offset(doc, offset):
+    l = 0
+    r = len(doc)
+    while r - l > 1:
+        mid = (l + r) / 2
+        if doc[mid].idx + len(doc[mid]) > offset:
+            r = mid
+        else:
+            l = mid
+
+    if doc[l].idx + len(doc[l]) == offset:
+        return doc[l]
+    else:
+        return None
+
+
+def find_overlapping(doc):
+    return filter(lambda c: any_overlapping_spans(c[0], c[1]),
+                  itertools.combinations(doc.get_entities(), 2))
+
+
+def any_overlapping_spans(a, b):
+        for i in a.spans:
+            for j in b.spans:
+                if j[0] < i[1] and i[0] < j[1]:
+                    return True
+        return False
+
+
+# Requires the spans be independent
+def spans_to_biluo(spans, total_len):
+    # o -> 0, b -> 1, i -> 2, l -> 3, u -> 4
+    spans = sorted(spans)
+    result = [0] * total_len
+    i = 0
+    j = 0
+    while j < len(spans):
+        if i >= total_len:
+            print spans
+        if i < spans[j][0]:  # o
+            result[i] = 0
+        elif i == spans[j][0] and (i + 1 < spans[j][1]):  # b
+            result[i] = 1
+        elif i == spans[j][0]:  # u
+            result[i] = 4
+            j += 1
+        elif i + 1 == spans[j][1]:  # l
+            result[i] = 3
+            j += 1
+        else:  # i
+            result[i] = 2
+        i += 1
+    return result
+
+
+def compatibilize_tokenization(original_docs):
+    boundaries = sorted(set(itertools.chain.from_iterable((
+        doc._get_entity_boundaries_for_tokenization() for doc in original_docs))))
```

Comment:
I will make the change as you suggest - it will indeed make life easier. I was operating under what I thought was an earlier discussion that we wanted to respect annotation boundaries despite what spacy would give us for tokenization.